Knowledge Distillation 
===============================

**Author**: [Clara Martinez](https://github.com/moonblume/LIVIA.git)


Knowledge distillation is a technique that enables knowledge transfer
from large, computationally expensive models to smaller ones without
losing validity. This allows for deployment on less powerful hardware,
making evaluation faster and more efficient.

Librairies
================


In [21]:
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "2"
import pandas as pd
import numpy as np

import torch
import torch.nn as nn
import torch.optim as optim
import torchvision.transforms as transforms
from torch.utils.data import DataLoader, TensorDataset
from torch.optim.lr_scheduler import ReduceLROnPlateau

from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from scipy.signal import savgol_filter
from sklearn.metrics import mean_absolute_error, mean_squared_error

from typing import List, Union, Tuple, Any
import statistics

# Check if GPU is available, and if not, use the CPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

Loading dataset
================

Fisrt, I focus on the physiological signals of the Biovid dataset. In one sample, we have access to 6 classes associated with 0 to 4 pain levels :  

Time: This could be the timestamp or time index when the signal was recorded.

GSR (Galvanic Skin Response): A measure of the electrical conductance of the skin, which varies with the moisture level of the skin. It's often associated with emotional arousal.

ECG (Electrocardiogram): A recording of the electrical activity of the heart over time. It typically consists of waves representing the depolarization and repolarization of the heart muscle during each heartbeat.

EMG (Electromyography) - Trapezius: Measures the electrical activity produced by skeletal muscles. The trapezius muscle is a large superficial muscle that extends longitudinally from the occipital bone to the lower thoracic vertebrae and laterally to the spine of the scapula.

EMG - Corrugator: Electromyography signal from the corrugator supercilii muscle, which is a small facial muscle involved in frowning and expressing negative emotions.

EMG - Zygomaticus: Electromyography signal from the zygomaticus major muscle, which is involved in smiling and expressing positive emotions.  
    
    
Our objective is to predict the pain level of input signals. One signal corresponds to one csv file.



In [22]:
# Define the directory containing the CSV files
biosignals_path = '/home/ens/AU59350/LIVIA/physio/physio_organised/'

# Initialize an empty list to store data for DataFrame
data = []

# Iterate over each pain level directory
for pain_level in os.listdir(biosignals_path):
    pain_level_dir = os.path.join(biosignals_path, pain_level)
    
    # Check if it's a directory
    if os.path.isdir(pain_level_dir):
        # Iterate over each CSV file in the pain level directory
        for csv_file in os.listdir(pain_level_dir):
            # Check if it's a CSV file
            if csv_file.endswith('.csv'):
                csv_path = os.path.join(pain_level_dir, csv_file)
                # Read the CSV file
                df = pd.read_csv(csv_path, sep='\t')
                # Extract GSR values
                gsr_signal = df['gsr'].values
                # Append the CSV name, GSR signals, and Pain level to the data list
                data.append({'CSV name': csv_file, 'GSR signals': gsr_signal, 'Pain level': int(pain_level)})

# Create a DataFrame from the collected data
df = pd.DataFrame(data)

# Display the DataFrame
df.head()


CSV name  \
0  072414_m_23-PA2-034_bio.csv   
1  081609_w_40-PA2-028_bio.csv   
2  081714_m_36-PA2-065_bio.csv   
3  102514_w_40-PA2-046_bio.csv   
4  120514_w_56-PA2-019_bio.csv   

                                         GSR signals  Pain level  
0  [6.966839, 6.966161, 6.966, 6.966839, 6.966161...           2  
1  [0.872, 0.872, 0.872, 0.872, 0.872, 0.872, 0.8...           2  
2  [6.089862, 6.091, 6.091432, 6.092, 6.092432, 6...           2  
3  [1.462, 1.462, 1.462, 1.462, 1.462, 1.462, 1.4...           2  
4  [2.226, 2.226, 2.226, 2.226, 2.226, 2.226, 2.2...           2

### Selection of number of pain level included in the classification task

In [23]:
# Filter the DataFrame to keep rows where pain level is not equal to 1, 2, or 3
df = df[~df['Pain level'].isin([1, 2, 3])]

# Print the filtered DataFrame
print(df)

                         CSV name  \
5220  080314_w_25-PA4-067_bio.csv   
5221  092009_m_54-PA4-042_bio.csv   
5222  071709_w_23-PA4-071_bio.csv   
5223  082809_m_26-PA4-005_bio.csv   
5224  112909_w_20-PA4-080_bio.csv   
...                           ...   
8695  082909_m_47-BL1-085_bio.csv   
8696  081609_w_40-BL1-090_bio.csv   
8697  091809_w_43-BL1-097_bio.csv   
8698  112016_m_25-BL1-091_bio.csv   
8699  083013_w_47-BL1-086_bio.csv   

                                            GSR signals  Pain level  
5220  [4.521143, 4.522, 4.522, 4.522, 4.522, 4.522, ...           4  
5221  [1.673, 1.673, 1.673, 1.673, 1.673, 1.673, 1.6...           4  
5222  [4.326707, 4.327, 4.326292, 4.326708, 4.327, 4...           4  
5223  [3.913801, 3.913, 3.913, 3.913, 3.912801, 3.91...           4  
5224  [2.633, 2.633728, 2.634, 2.633272, 2.633728, 2...           4  
...                                                 ...         ...  
8695  [3.184, 3.184399, 3.1846, 3.184, 3.184, 3.184,...          

Preprocessing
================

Preprocessing steps for GSR DataFrame include tasks such as handling missing values, smoothing the signal to reduce noise in the GSR signal 9(Savitzky-Golay filtering), removing outliers (z-score), and normalizing the data between a specified range, such as [0, 1] or [-1, 1] helping comparison across different subjects.

In [24]:
# Function to preprocess GSR signals
def preprocess_gsr_signal(gsr_signal):
    # Handle missing values (if any)
    gsr_signal = np.array(gsr_signal)  # Convert to NumPy array
    gsr_signal = gsr_signal[~np.isnan(gsr_signal)]  # Remove NaN values
    
    # Check if the length of the signal is sufficient for smoothing
    if len(gsr_signal) < 5:
        # If the signal is too short, return the original signal
        return gsr_signal
    
    try:
        # Smoothing using Savitzky-Golay filter
        gsr_signal_smooth = savgol_filter(gsr_signal, window_length=5, polyorder=2)
    except ValueError:
        # If an error occurs during smoothing, return the original signal
        return gsr_signal
    
    # Removing outliers based on Z-scores
    z_scores = (gsr_signal_smooth - gsr_signal_smooth.mean()) / gsr_signal_smooth.std()
    gsr_signal_smooth_no_outliers = gsr_signal_smooth[(z_scores < 3)]
    
    # Normalization
    if len(gsr_signal_smooth_no_outliers) > 0:
        gsr_signal_normalized = (gsr_signal_smooth_no_outliers - gsr_signal_smooth_no_outliers.min()) / \
                                 (gsr_signal_smooth_no_outliers.max() - gsr_signal_smooth_no_outliers.min())
    else:
        # If there are no valid values after removing outliers, return the original signal
        return gsr_signal
    
    return gsr_signal_normalized

# Apply preprocessing to each row in the DataFrame
df['GSR signals'] = df['GSR signals'].apply(preprocess_gsr_signal)

# Display the updated DataFrame
df.head()

CSV name  \
5220  080314_w_25-PA4-067_bio.csv   
5221  092009_m_54-PA4-042_bio.csv   
5222  071709_w_23-PA4-071_bio.csv   
5223  082809_m_26-PA4-005_bio.csv   
5224  112909_w_20-PA4-080_bio.csv   

                                            GSR signals  Pain level  
5220  [0.0, 0.000784955924601847, 0.0012131137016623...           4  
5221  [0.21654547886192313, 0.21654547886192313, 0.2...           4  
5222  [1.0, 0.9983642739356123, 0.9979709555643436, ...           4  
5223  [0.3629402970779174, 0.36178416384520784, 0.36...           4  
5224  [0.16994520796566095, 0.1741950993228739, 0.17...           4

In [27]:
print(df)

                         CSV name  \
5220  080314_w_25-PA4-067_bio.csv   
5221  092009_m_54-PA4-042_bio.csv   
5222  071709_w_23-PA4-071_bio.csv   
5223  082809_m_26-PA4-005_bio.csv   
5224  112909_w_20-PA4-080_bio.csv   
...                           ...   
8695  082909_m_47-BL1-085_bio.csv   
8696  081609_w_40-BL1-090_bio.csv   
8697  091809_w_43-BL1-097_bio.csv   
8698  112016_m_25-BL1-091_bio.csv   
8699  083013_w_47-BL1-086_bio.csv   

                                            GSR signals  Pain level  
5220  [0.0, 0.000784955924601847, 0.0012131137016623...           4  
5221  [0.21654547886192313, 0.21654547886192313, 0.2...           4  
5222  [1.0, 0.9983642739356123, 0.9979709555643436, ...           4  
5223  [0.3629402970779174, 0.36178416384520784, 0.36...           4  
5224  [0.16994520796566095, 0.1741950993228739, 0.17...           4  
...                                                 ...         ...  
8695  [0.9952579796341273, 0.9990653940310847, 1.0, ...          

In [28]:
import torch
from sklearn.model_selection import train_test_split

# Step 1: Preprocess GSR signals
def preprocess_gsr_signal(gsr_signal):
    # Your preprocessing code here
    return gsr_signal

df['GSR signals'] = df['GSR signals'].apply(preprocess_gsr_signal)

# Step 2: Pad or truncate signals
max_length = max(len(signal) for signal in df['GSR signals'])
gsr_signals = np.array([np.pad(signal, (0, max_length - len(signal))) if len(signal) < max_length else signal[:max_length] for signal in df['GSR signals']])

# Step 3: Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(gsr_signals, df['Pain level'].values, test_size=0.2, random_state=42)

# Step 4: Convert data into PyTorch tensors
X_train_tensor = torch.tensor(X_train, dtype=torch.float32).unsqueeze(1)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32).unsqueeze(1)
y_train_tensor = torch.tensor(y_train, dtype=torch.long)  # Assuming your labels are integer-encoded
y_test_tensor = torch.tensor(y_test, dtype=torch.long)

# Check the shape of tensors
print("X_train_tensor shape:", X_train_tensor.shape)
print("X_test_tensor shape:", X_test_tensor.shape)
print("y_train_tensor shape:", y_train_tensor.shape)
print("y_test_tensor shape:", y_test_tensor.shape)


X_train_tensor shape: torch.Size([2784, 1, 2816])
X_test_tensor shape: torch.Size([696, 1, 2816])
y_train_tensor shape: torch.Size([2784])
y_test_tensor shape: torch.Size([696])


In [25]:
# Step 1: Prepare the data
max_length = max(len(signal) for signal in df['GSR signals'])  # Find the maximum length of GSR signals

# Pad or truncate the GSR signals to the maximum length
gsr_signals = np.array([np.pad(signal, (0, max_length - len(signal))) if len(signal) < max_length else signal[:max_length] for signal in df['GSR signals']])

pain_levels = df['Pain level'].values

# Step 2: Split the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(gsr_signals, pain_levels, test_size=0.2, random_state=42)

# Step 3: Convert the data into PyTorch tensors
X_train_tensor = torch.tensor(X_train, dtype=torch.float32).unsqueeze(1)  
X_test_tensor = torch.tensor(X_test, dtype=torch.float32).unsqueeze(1)  
y_train_tensor = torch.tensor(y_train, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.float32)

# Check the shape of tensors
print("X_train_tensor shape:", X_train_tensor.shape)
print("X_test_tensor shape:", X_test_tensor.shape)
print("y_train_tensor shape:", y_train_tensor.shape)
print("y_test_tensor shape:", y_test_tensor.shape)


X_train_tensor shape: torch.Size([2784, 1, 2816])
X_test_tensor shape: torch.Size([696, 1, 2816])
y_train_tensor shape: torch.Size([2784])
y_test_tensor shape: torch.Size([696])


X_train_tensor: The training data tensor with a shape of [6960, 1, 2816], indicating that there are 6960 samples, each with 1 channel (for the GSR signal), and each signal has been padded or truncated to a length of 2816.  
X_test_tensor: The test data tensor with a shape of [1740, 1, 2816], indicating that there are 1740 samples in the test set, each with 1 channel, and the signals have the same length as the training data.  
y_train_tensor: The training labels tensor with a shape of [6960], containing the corresponding pain levels for the training samples.  
y_test_tensor: The test labels tensor with a shape of [1740], containing the corresponding pain levels for the test samples.

The first dimension (29) represents the batch size, indicating that there are 29 samples in the batch.  
The second dimension (1) represents the number of channels. In this case, there is only one channel.  
The third dimension (1) represents the length of the input data for each channel.

Input data has a shape of (batch_size, channels, sequence_length).

Defining model classes and utility functions
================

In [8]:
# Neural network class to be used as teacher:

class Conv1D_T(nn.Module):
    def __init__(self, num_classes=2):
        super(Conv1D_T, self).__init__()
        # First Convolutional Layer
        self.conv1 = nn.Conv1d(in_channels=1, out_channels=32, kernel_size=5, stride=2)
        self.relu1 = nn.ReLU()
        self.maxpool1 = nn.MaxPool1d(kernel_size=2)
        
        # Second Convolutional Layer
        self.conv2 = nn.Conv1d(in_channels=32, out_channels=64, kernel_size=5)
        self.relu2 = nn.ReLU()
        self.maxpool2 = nn.MaxPool1d(kernel_size=2)
        
        # Fully Connected Layers
        self.fc1 = nn.Linear(22336, 512)  
        self.relu3 = nn.ReLU()
        self.fc2 = nn.Linear(512, num_classes)
        
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.conv1(x)
        x = self.relu1(x)
        x = self.maxpool1(x)
        
        x = self.conv2(x)
        x = self.relu2(x)
        x = self.maxpool2(x)
        
        x = x.view(x.size(0), -1)  # Flatten the tensor to 1D
        x = self.fc1(x)
        
        x = self.fc2(x)
        #x = self.sigmoid(x)
        return x

# Neural network class to be used as student:

class Conv1D_S(nn.Module):
    def __init__(self, num_classes=2):
        super(Conv1D_S, self).__init__()
        # First Convolutional Layer
        self.conv1 = nn.Conv1d(in_channels=1, out_channels=32, kernel_size=5, stride=2)
        self.relu1 = nn.ReLU()
        self.maxpool1 = nn.MaxPool1d(kernel_size=2)
        
        # Second Convolutional Layer
        self.conv2 = nn.Conv1d(in_channels=32, out_channels=64, kernel_size=5)
        self.relu2 = nn.ReLU()
        self.maxpool2 = nn.MaxPool1d(kernel_size=2)
        
        # Fully Connected Layers
        self.fc1 = nn.Linear(22336, 512)  
        self.relu3 = nn.ReLU()
        self.fc2 = nn.Linear(512, num_classes)
        
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        x = self.conv1(x)
        x = self.relu1(x)
        x = self.maxpool1(x)
        
        x = self.conv2(x)
        x = self.relu2(x)
        x = self.maxpool2(x)
        
        x = x.view(x.size(0), -1)  # Flatten the tensor to 1D
        x = self.fc1(x)
        
        x = self.fc2(x)
        #x = self.sigmoid(x)
        return x


In [20]:
def train(model, train_loader, epochs, learning_rate, device):
    # Define learning parameters
    learning_rate = 0.0001
    epochs = 100
    batch_size = 1024
    num_classes = 2

    # Define loss function and optimizer
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(), lr=learning_rate, momentum=0.9)

    # Initialize K-Fold cross-validation
    k_folds = 5
    kf = KFold(n_splits=k_folds, shuffle=True)
    
    # Perform cross-validation
    for fold, (train_index, val_index) in enumerate(kf.split(X_train_tensor)):
        print(f"Fold {fold + 1}/{k_folds}")
        
        # Split data into training and validation sets for this fold
        X_train_fold, X_val_fold = X_train_tensor[train_index], X_train_tensor[val_index]
        y_train_fold, y_val_fold = y_train_tensor[train_index], y_train_tensor[val_index]
        
        # Create DataLoader for training and validation sets
        train_dataset = TensorDataset(X_train_fold, y_train_fold)
        val_dataset = TensorDataset(X_val_fold, y_val_fold)
        train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
        val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
        
        for epoch in range(epochs):
            
            model.train()
            running_loss = 0.0
            correct = 0
            total = 0

            for inputs, labels in train_loader:
                optimizer.zero_grad()
                outputs = model(inputs)
                loss = criterion(outputs, labels)
                loss.backward()
                optimizer.step()
                running_loss += loss.item() * inputs.size(0)

                # Calculate training accuracy
                _, predicted = torch.max(outputs, 1)
                total += labels.size(0)
                correct += (predicted == labels).sum().item()

            epoch_loss = running_loss / len(train_loader.dataset)
            epoch_accuracy = correct / total
            print(f"Epoch {epoch+1}/{epochs}, Loss: {epoch_loss:.4f}, Accuracy: {epoch_accuracy:.4f}")

def test(model, test_loader, device):
    model.to(device)
    model.eval()

    correct = 0
    total = 0

    with torch.no_grad():
        outputs = model(X_test_tensor)
        predictions = torch.argmax(outputs, dim=1)
        y_test_tensor_sampled = y_test_tensor[:len(predictions)]
        accuracy = torch.mean((predictions == y_test_tensor_sampled).float()).item()
        print(f"Test Accuracy: {accuracy:.4f}")
        
        # Calculate MAE and RMSE
        mae = mean_absolute_error(y_test_tensor_sampled, predictions)
        rmse = mean_squared_error(y_test_tensor_sampled, predictions, squared=False)
        print(f"MAE: {mae:.4f}")
        print(f"RMSE: {rmse:.4f}")

Cross-entropy runs
==================

For reproducibility, we need to set the torch manual seed. I train
networks using different methods, so to compare them fairly, it makes
sense to initialize the networks with the same weights. I start by
training the teacher network using cross-entropy:

In [18]:
# Define batch size
batch_size = 1024

# Define training and testing datasets
train_dataset = TensorDataset(X_train_tensor, y_train_tensor)
test_dataset = TensorDataset(X_test_tensor, y_test_tensor)

# Create DataLoader for training and testing sets
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

# Train the teacher network
torch.manual_seed(42)
nn_t = Conv1D_T(num_classes=2).to(device)
train(nn_t, train_loader, epochs=100, learning_rate=0.0001, device=device)

# Test the teacher network
test_accuracy_T = test(nn_t, test_loader, device)

# Instantiate the lightweight network
torch.manual_seed(42)
nn_s = Conv1D_S(num_classes=2).to(device)


Fold 1/5
Epoch 1/100, Loss: 0.6974, Accuracy: 0.4946
Epoch 2/100, Loss: 0.6942, Accuracy: 0.4932
Epoch 3/100, Loss: 0.6941, Accuracy: 0.5068
Epoch 4/100, Loss: 0.6943, Accuracy: 0.5041
Epoch 5/100, Loss: 0.6933, Accuracy: 0.5156
Epoch 6/100, Loss: 0.6931, Accuracy: 0.5009
Epoch 7/100, Loss: 0.6933, Accuracy: 0.5063
Epoch 8/100, Loss: 0.6931, Accuracy: 0.5022
Epoch 9/100, Loss: 0.6929, Accuracy: 0.5072
Epoch 10/100, Loss: 0.6929, Accuracy: 0.5153
Epoch 11/100, Loss: 0.6928, Accuracy: 0.5165
Epoch 12/100, Loss: 0.6926, Accuracy: 0.5138
Epoch 13/100, Loss: 0.6926, Accuracy: 0.5084
Epoch 14/100, Loss: 0.6925, Accuracy: 0.5074
Epoch 15/100, Loss: 0.6924, Accuracy: 0.5154
Epoch 16/100, Loss: 0.6923, Accuracy: 0.5174
Epoch 17/100, Loss: 0.6922, Accuracy: 0.5183
Epoch 18/100, Loss: 0.6921, Accuracy: 0.5198
Epoch 19/100, Loss: 0.6921, Accuracy: 0.5217
Epoch 20/100, Loss: 0.6920, Accuracy: 0.5207
Epoch 21/100, Loss: 0.6919, Accuracy: 0.5198
Epoch 22/100, Loss: 0.6918, Accuracy: 0.5208
Epoch 23/1

/home/ens/AU59350/anaconda3/envs/ClaraEnv/lib/python3.10/site-packages/sklearn/metrics/_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


I instantiate one more lightweight network model to compare their
performances. Back propagation is sensitive to weight initialization, so
I need to make sure these two networks have the exact same
initialization.


In [ ]:
torch.manual_seed(42)
new_Conv1D_S = Conv1D_S(num_classes=2).to(device)

To ensure I have created a copy of the first network, we inspect the
norm of its first layer. If it matches, then the networks are indeed the same.

In [ ]:
# Print the norm of the first layer of the initial lightweight model
print("Norm of 1st layer of nn_light:", torch.norm(Conv1D_S.features[0].weight).item())
# Print the norm of the first layer of the new lightweight model
print("Norm of 1st layer of new_nn_light:", torch.norm(new_Conv1D_S.features[0].weight).item())

Print the total number of parameters in each model:

In [ ]:
total_params_T = "{:,}".format(sum(p.numel() for p in Conv1D_T.parameters()))
print(f"DeepNN parameters: {total_params_T}")
total_params_S = "{:,}".format(sum(p.numel() for p in Conv1D_S.parameters()))
print(f"LightNN parameters: {total_params_S}")

Train and test the lightweight network with cross entropy loss:

In [ ]:
train(Conv1D_S, train_loader, epochs=10, learning_rate=0.0001, device=device)
test_accuracy_S_ce = test(Conv1D_S, test_loader, device)

As we can see, based on test accuracy, I can now compare the deeper
network that is to be used as a teacher with the lightweight network
that is the supposed student. So far, the student has not intervened
with the teacher, therefore this performance is achieved by the student
itself. The metrics so far can be seen with the following lines:

In [ ]:
print(f"Teacher accuracy: {test_accuracy_T:.2f}%")
print(f"Student accuracy: {test_accuracy_S_ce:.2f}%")

In [ ]:
def train_knowledge_distillation(teacher, student, train_loader, epochs, learning_rate, T, soft_target_loss_weight, ce_loss_weight, device):
    # Define learning parameters
    learning_rate = 0.0001

    # Define loss function and optimizer
    ce_loss = nn.CrossEntropyLoss()
    optimizer = optim.SGD(student.parameters(), lr=learning_rate, momentum=0.9)

    teacher.eval()  # Teacher set to evaluation mode
    student.train() # Student to train mode

    for epoch in range(epochs):
        running_loss = 0.0
        for inputs, labels in train_loader:
            inputs, labels = inputs.to(device), labels.to(device)

            optimizer.zero_grad()

            # Forward pass with the teacher model - do not save gradients here as we do not change the teacher's weights
            with torch.no_grad():
                teacher_logits = teacher(inputs)

            # Forward pass with the student model
            student_logits = student(inputs)

            #Soften the student logits by applying softmax first and log() second
            soft_targets = nn.functional.softmax(teacher_logits / T, dim=-1)
            soft_prob = nn.functional.log_softmax(student_logits / T, dim=-1)

            # Calculate the soft targets loss. Scaled by T**2 as suggested by the authors of the paper "Distilling the knowledge in a neural network"
            soft_targets_loss = torch.sum(soft_targets * (soft_targets.log() - soft_prob)) / soft_prob.size()[0] * (T**2)

            # Calculate the true label loss
            label_loss = ce_loss(student_logits, labels)

            # Weighted sum of the two losses
            loss = soft_target_loss_weight * soft_targets_loss + ce_loss_weight * label_loss

            loss.backward()
            optimizer.step()

            running_loss += loss.item()

        print(f"Epoch {epoch+1}/{epochs}, Loss: {running_loss / len(train_loader)}")

# Apply ``train_knowledge_distillation`` with a temperature of 2. Arbitrarily set the weights to 0.75 for CE and 0.25 for distillation loss.
train_knowledge_distillation(teacher=Conv1D_T, student=new_Conv1D_S, train_loader=train_loader, epochs=10, learning_rate=0.001, T=2, soft_target_loss_weight=0.25, ce_loss_weight=0.75, device=device)
test_accuracy_light_ce_and_kd = test(new_nn_light, test_loader, device)

# Compare the student test accuracy with and without the teacher, after distillation
print(f"Teacher accuracy: {test_accuracy_T:.2f}%")
print(f"Student accuracy without teacher: {test_accuracy_S_ce:.2f}%")
print(f"Student accuracy with CE + KD: {test_accuracy_light_ce_and_kd:.2f}%")